File: S02/week12/sejun/ValidParentheses.java
// 1ms  41.42MB

class Solution {
    private Stack<Character> left = new Stack<>();

    public boolean isValid(String s) {
        for (char c : s.toCharArray()) {
            if (c == '[' || c == '{' || c == '(') {
                left.add(c);
            } else {
                if (left.isEmpty() || !(Math.abs(c - left.pop()) <= 2)) {
                    return false;
                }
            }

        }

        return left.isEmpty();
    }
}

File: S02/week12/seokjoon/ValidParentheses.kt
//164ms, 38.10mb

class Solution {
    fun isValid(s: String): Boolean {
        val stack : Stack<Char> = Stack() 

        s.forEach {
            if(setOf('(','{','[').contains(it)){
                stack.push(it)
            }else {
                if(stack.isEmpty()) return false
                when(it){
                    ')' -> if(stack.pop() !='(') return false
                    '}' -> if(stack.pop() !='{') return false
                    else -> if(stack.pop() !='[') return false
                }
            }
        }
        return stack.isEmpty()
    }
}
